src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+    Add(GlobalId),
+    Drop(GlobalId),
+    Resume(GlobalId, Trace),
+    AllowCompaction(GlobalId, Timestamp),
+}
+
+/// A Batch contains all of the updates that originated within some time range [lower, upper)
+/// but the data live the file at `path`.
+///
+/// The data stored in each batch are triples of (Row, Timestamp, Diff) such that there is
+/// exactly one copy of each (Row, Timestamp) in each batch. Batches also have a header
+/// and a footer indicating the upper and lower bound timestamps.
+/// TODO: Batches are meant to mimic differential / cdcv2 batches but do not do so
+/// currently. Let's fix that. Specifically, introduce a `since` field, and counts
+/// for the number of updates at each time.
+#[derive(Debug)]
+struct Batch {
+    upper: Timestamp,
+    lower: Timestamp,
+    path: PathBuf,
+}
+
+impl Batch {
+    /// Create a batch from a finished log segment file.
+    fn create(log_segment_path: &Path, trace_path: &Path) -> Result<Self, anyhow::Error> {
+        let messages = read_segment(log_segment_path)?;
+        Batch::create_from_messages(messages, trace_path, None)
+    }
+
+    /// Read in and consolidate a list of messages, and write them to a new batch file
+    /// in `trace_path`.
+    ///
+    /// Will also compact updates up to `compaction_frontier` if provided..
+    fn create_from_messages(
+        messages: Vec<Message>,
+        trace_path: &Path,
+        compaction_frontier: Option<Timestamp>,
+    ) -> Result<Self, anyhow::Error> {
+        let mut upper: Option<Timestamp> = None;
+        let mut lower: Option<Timestamp> = None;
+        let mut time_data = BTreeMap::new();
+        for message in messages.iter() {
+            match message {
+                Message::Progress(time) => match (lower, upper) {
+                    (None, None) => {
+                        lower = Some(*time);
+                    }
+                    (Some(l), None) => {
+                        assert!(*time >= l);
+                        upper = Some(*time);
+                    }
+                    (Some(_), Some(u)) => {
+                        assert!(*time >= u);
+                        upper = Some(*time);
+                    }
+                    (None, Some(_)) => unreachable!(),
+                },
+                Message::Data(Update {
+                    row,
+                    timestamp,
+                    diff,
+                }) => {
+                    let time = if let Some(frontier) = compaction_frontier {
+                        std::cmp::min(frontier, *timestamp)
+                    } else {
+                        *timestamp
+                    };
+                    let entry = time_data.entry((time, row)).or_insert(0);
+                    *entry += diff;
+                }
+            }
+        }
+
+        // Now let's prepare the output
+        let mut buf = Vec::new();
+        assert!(lower.is_some());
+        assert!(upper.is_some());
```

Comment:
It feels like by this point there need to have been some invariants about the log file. That's fine, but locally I don't know what they are. There should be at least two `Progress` messages, for example, to set both `lower` and `upper`. Could be legit, but I want to call out that this is locally difficult to know / verify.